akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala

```diff
@@ -0,0 +1,130 @@
+   * in PEM as "PRIVATE KEY").
+   * @throws PEMLoadingException when the `derData` is for an unsupported format
+   */
+  @ApiMayChange
+  @throws[PEMLoadingException]("when the `derData` is for an unsupported format")
+  def load(derData: DERData): PrivateKey = {
+    derData.label match {
+      case "RSA PRIVATE KEY" =>
+        loadPkcs1PrivateKey(derData.bytes)
+      case "PRIVATE KEY" =>
+        loadPkcs8PrivateKey(derData.bytes)
+      case unknown =>
+        throw new PEMLoadingException(s"Don't know how to read a private key from PEM data with label [$unknown]")
+    }
+  }
+
+  private def loadPkcs1PrivateKey(bytes: Array[Byte]) = {
+    val derInputStream = new ASN1InputStream(new DERDecoder, bytes)
+    // Here's the specification: https://tools.ietf.org/html/rfc3447#appendix-A.1.2
+    val sequence = {
+      try {
+        derInputStream.readObject[ASN1Sequence]()
+      } finally {
+        derInputStream.close()
+      }
+    }
+    val version = getInteger(sequence, 0, "version").intValueExact()
+    if (version < 0 || version > 1) {
+      throw new IllegalArgumentException(s"Unsupported PKCS1 version: $version")
+    }
+    val modulus = getInteger(sequence, 1, "modulus")
+    val publicExponent = getInteger(sequence, 2, "publicExponent")
+    val privateExponent = getInteger(sequence, 3, "privateExponent")
+    val prime1 = getInteger(sequence, 4, "prime1")
+    val prime2 = getInteger(sequence, 5, "prime2")
+    val exponent1 = getInteger(sequence, 6, "exponent1")
+    val exponent2 = getInteger(sequence, 7, "exponent2")
+    val coefficient = getInteger(sequence, 8, "coefficient")
+
+    val keySpec = if (version == 0) {
+      new RSAPrivateCrtKeySpec(
+        modulus,
+        publicExponent,
+        privateExponent,
+        prime1,
+        prime2,
+        exponent1,
+        exponent2,
+        coefficient)
+    } else {
+      // Does anyone even use multi-primes? Who knows, maybe this code will never be used. Anyway, I guess it will work,
+      // the spec isn't exactly complicated.
+      val otherPrimeInfosSequence = getSequence(sequence, 9, "otherPrimeInfos")
+      val otherPrimeInfos = (for (i <- 0 until otherPrimeInfosSequence.size()) yield {
+        val name = s"otherPrimeInfos[$i]"
+        val seq = getSequence(otherPrimeInfosSequence, i, name)
+        val prime = getInteger(seq, 0, s"$name.prime")
+        val exponent = getInteger(seq, 1, s"$name.exponent")
+        val coefficient = getInteger(seq, 2, s"$name.coefficient")
+        new RSAOtherPrimeInfo(prime, exponent, coefficient)
+      }).toArray
+      new RSAMultiPrimePrivateCrtKeySpec(
+        modulus,
+        publicExponent,
+        privateExponent,
+        prime1,
+        prime2,
+        exponent1,
+        exponent2,
+        coefficient,
+        otherPrimeInfos)
+    }
+
+    val keyFactory = KeyFactory.getInstance("RSA")
```

Comment:
I assume it's safe to always generate RSA keys.

Reply:
This is reading PKCS#1 si it must be RSA.

Reply:
I am paranoid so I checked [the PKCS#8 case too](https://github.com/openjdk/jdk/blob/6bab0f539fba8fb441697846347597b4a0ade428/src/java.base/share/classes/sun/security/rsa/RSAKeyFactory.java#L38-L66).